metricbeat/module/coredns/stats/stats.go

```diff
@@ -24,16 +24,27 @@ import (
 
 var mapping = &prometheus.MetricsMapping{
 	Metrics: map[string]prometheus.MetricMap{
-		"coredns_panic_count_total":       prometheus.Metric("panic.count.total"),
-		"coredns_dns_request_count_total": prometheus.Metric("dns.request.count.total"),
+
+		// base CoreDNS metrics
+		"coredns_panic_count_total":       prometheus.Metric("panic.count"),
+		"coredns_dns_request_count_total": prometheus.Metric("dns.request.count"),
 		"coredns_dns_request_duration_seconds": prometheus.Metric(
 			"dns.request.duration.ns",
 			prometheus.OpMultiplyBuckets(1000000000)),
 		"coredns_dns_request_size_bytes":         prometheus.Metric("dns.request.size.bytes"),
-		"coredns_dns_request_do_count_total":     prometheus.Metric("dns.request.do.count.total"),
-		"coredns_dns_request_type_count_total":   prometheus.Metric("dns.request.type.count.total"),
+		"coredns_dns_request_do_count_total":     prometheus.Metric("dns.request.do.count"),
+		"coredns_dns_request_type_count_total":   prometheus.Metric("dns.request.type.count"),
 		"coredns_dns_response_size_bytes":        prometheus.Metric("dns.response.size.bytes"),
-		"coredns_dns_response_rcode_count_total": prometheus.Metric("dns.response.rcode.count.total"),
+		"coredns_dns_response_rcode_count_total": prometheus.Metric("dns.response.rcode.count"),
+
+		// cache plugin metrics (might not be present if cache plugin is not configured)
+		"coredns_cache_hits_total":   prometheus.Metric("dns.cache.hits.count"),
+		"coredns_cache_misses_total": prometheus.Metric("dns.cache.misses.count"),
+
+		// go runtime
```

Comment:
As we expect these also in other places, I wonder if we should introduce `runtime` on the root level instead.

Reply:
not sure if I'm on the same page.

Right now event path is `coredns.stats` which makes the event look like
```
    {
        "coredns": {
            "stats": {
                "panic": {
                    "count": 0
                },
                "runtime": {
                    "go": {
                        "routines": {
                            "count": 21
                        }
                    },
                    "memory": {
                        "alloc": {
                            "bytes": 1843744
                        }
                    },
                    "process": {
                        "threads": {
                            "count": 18
                        }
                    }
                }
            }
        },
        "event": {
            "dataset": "coredns.stats",
            "duration": 115000,
            "module": "coredns"
        },
        "metricset": {
            "name": "stats"
        },
        "service": {
            "address": "127.0.0.1:55555",
            "type": "coredns"
        }
    },
```

do you registering creating a new metricset by the name `coredns.runtime` ? and parsing only the runtime data?

Reply:
I agree `runtime` should be in the root level of the event, that means:

```
    {
      "runtime": {
        "go": {
          "routines": {
            "count": 21
          },
          ...
        },
        "coredns": {
            "stats": {
                "panic": {
                    "count": 0
                },
            }
        },
        ...
    },
```

That said, I think the Prometheus helper framework doesn't support this? We have done tricks like this in the past: https://github.com/elastic/beats/blob/24994492d669b43d6c89db9737016b21ab888215/metricbeat/module/kubernetes/state_pod/state_pod.go#L109-L117 but I would prefer to keep them to a minimum, we should implement a better way to get this.

Also, getting `runtime` metrics from the `stats` metricset may sound wrong? It could make sense to add a `runtime` metricset, common to all modules supporting this, to make things uniform.

Perhaps we should move `runtime` metrics to a different issue to unblock this, then add support for them the right way.

Reply:
about having a different `metricset`, I take that back, as that would mean double requesting the same info from the exporter. I'm ok with this :+1:

About moving things to the root level, do you foresee this becoming an ECS field @ruflin?

Reply:
Yes for ECS. I remember we had somewhere else a discussion about this with @webmat .

For having things on the root level: Can we extend the prometheus helper to make this possible? I assume now that all is based on the reporter interface this should be easier.

My take to unblock this PR would be to add for now `runtime` inside the metricset with all the prefixes and abstract it out later. Not optimal but keeps us moving.

Reply:
@exekias Interesting question if 1 request should be able to create 2 events or if metricsets are ok. To keep complexity low I would argue having 2 requests for the same thing is ok.

Reply:
abstracting it out later would imply deprecating these metrics in favor of the new names, is that correct? If we are discussing adding these for all services I think they deserve a better solution, so we can share mapping. How about leaving them out of this PR and open an issue to tackle runtime metrics with prometheus helper?

Reply:
@ruflin @exekias 
I removed runtime metrics before reading these new comments.

If you are ok I'll move forward, `runtime` should come from here https://github.com/elastic/beats/issues/11836

I don't think it will take us very long to come to an agreement, and then I will add back the metrics. In the meantime I'll open an issue for adding the root namespaced items at the prometheus helper.

Is that ok with you?

Reply:
I like the idea of a `runtime` fieldset, please ping me on the PR whenever you work on it 👀 @odacremolbap